File: sources/Driver.cpp
#include "../includes/Driver.hpp"
#include "../includes/Stats.hpp"
#include <iostream>


Driver::Driver(const std::string &_Name, const std::string &_Team, TireType tireType, Weather weather) 
    : _Name(_Name), _Team(_Team), _Position(0), _CurrentGear("0"), _Lap(1), _TileIndex(0), _SquareIndex(0), _LaneIndex(0), _StartingTile(0), _Tire(tireType, weather) {}
    
void Driver::setPosition(int pos) { _Position = pos; }
void Driver::setCurrentGear(const std::string& gear) { _CurrentGear = gear; }
void Driver::setTileIndex(const int index) { _TileIndex = index; };
void Driver::setLaneIndex(const int index) { _LaneIndex = index; };
void Driver::setSquareIndex(const int index) { _SquareIndex = index; };
void Driver::setInsideIndex(const int index) { _InsideIndex = index; };
void Driver::setStartingTile(const int index) { _StartingTile = index; };
void Driver::setLap(const int num) { _Lap = num; };
void Driver::setTire(TireType type, Weather weather) { _Tire = Tire(type, weather); };

std::string Driver::getName() const { return _Name; };
std::string Driver::getTeam() const { return _Team; };
int Driver::getPosition() const { return _Position; };
std::string Driver::getCurrentGear() const {return _CurrentGear; };
int Driver::getLap() const { return _Lap; };
Stats& Driver::getStats() { return _Stats; };
int Driver::getTileIndex() const { return _TileIndex; }
int Driver::getSquareIndex() const { return _SquareIndex; }
int Driver::getLaneIndex() const { return _LaneIndex; }
int Driver::getInsideIndex() const { return _InsideIndex; };

void Driver::moveForward(const Track& track)
{
    const Tile& currentTile = track.getTile(_TileIndex);
    auto it = currentTile.laneSquares.find(_LaneIndex);
    if (it == currentTile.laneSquares.end() || it->second.empty())
    {
        std::cerr << "Error: Driver is in a non-existing or empty lane!\n";
        return;
    }
    const std::vector<Square>& currentLaneSquares = it->second;
    if ((size_t)_SquareIndex + 1 < currentLaneSquares.size())
    {
        _SquareIndex++;
    }
    else
    {
        _TileIndex++;
        if (_TileIndex >= track.getTrackLength())
        {
            _Lap++;
            _TileIndex = 0;
        }
        const Tile& nextTile = track.getTile(_TileIndex);
        auto nextLaneIt = nextTile.laneSquares.find(_LaneIndex);
        if (nextLaneIt == nextTile.laneSquares.end() || nextLaneIt->second.empty())
        {
            std::cerr << "Error: Lane " << _LaneIndex << " does not exist in tile " << _TileIndex << "!\n";
            return;
        }
        _SquareIndex = 0;
        std::cout << _Name << " moved to tile " << _TileIndex << " (" << nextTile.color << ") in lane " << _LaneIndex << "\n";
    }
}

void Driver::changeLane(const Track& track)
{
    const Tile& currentTile = track.getTile(_TileIndex);
    while (true)
    {
        int direction;
        std::cout << "Enter lane change direction (1: Left, -1: Right): ";
        std::cin >> direction;
        int newLane = _LaneIndex + direction;
        std::cout << "new lane: " << newLane << std::endl;
        if (newLane >= 0 && newLane <= currentTile.lanes)
        {
            _LaneIndex = newLane;
            std::cout << _Name << " switched to lane " << _LaneIndex << "\n";
            break ;
        }
        else
        {
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            std::cout << "Invalid lane! Try again.\n";
        }
    }
}

void changeLaneOrMove(Track &track, Driver &driver)
{
    int action;
    while (true) 
    {
        std::cout << "Choose action (1: Move Forward, 2: Change Lane): ";
        std::cin >> action;

        if (std::cin.fail() || (action != 1 && action != 2))
        {
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            std::cout << "Invalid action! Try again.\n";
        }
        else
            break;
    }
    if (action == 1)
        driver.moveForward(track); 
    else if (action == 2)
        driver.changeLane(track);
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
}

File: sources/Driver.ts

import { Stats } from "./Stats.ts";
import { Tire, TireType, Weather } from "./Tire.ts";
import { Track } from "./Track.ts";

export class Driver {
  private name: string;
  private team: string;
  private position = 0;
  private currentGear = "0";
  private lap = 1;
  private tileIndex = 0;
  private squareIndex = 0;
  private laneIndex = 0;
  private startingTile = 0;
  private insideIndex = 0;
  private tire: Tire;
  private stats = new Stats();

  constructor(name: string, team: string, tireType: TireType, weather: Weather) {
    this.name = name;
    this.team = team;
    this.tire = new Tire(tireType, weather);
  }

  // Setters
  setPosition(pos: number): void { this.position = pos; }
  setCurrentGear(gear: string): void { this.currentGear = gear; }
  setTileIndex(index: number): void { this.tileIndex = index; }
  setLaneIndex(index: number): void { this.laneIndex = index; }
  setSquareIndex(index: number): void { this.squareIndex = index; }
  setInsideIndex(index: number): void { this.insideIndex = index; }
  setStartingTile(index: number): void { this.startingTile = index; }
  setLap(num: number): void { this.lap = num; }
  setTire(type: TireType, weather: Weather): void { this.tire = new Tire(type, weather); }

  // Getters
  getName(): string { return this.name; }
  getTeam(): string { return this.team; }
  getPosition(): number { return this.position; }
  getCurrentGear(): string { return this.currentGear; }
  getLap(): number { return this.lap; }
  getStats(): Stats { return this.stats; }
  getTileIndex(): number { return this.tileIndex; }
  getSquareIndex(): number { return this.squareIndex; }
  getLaneIndex(): number { return this.laneIndex; }
  getInsideIndex(): number { return this.insideIndex; }

  moveForward(track: Track): void {
    const currentTile = track.getTile(this.tileIndex);
    const currentLane = currentTile.laneSquares[this.laneIndex];
    if (!currentLane || currentLane.length === 0) {
      console.error("Error: Driver is in a non-existing or empty lane!");
      return;
    }

    if (this.squareIndex + 1 < currentLane.length) {
      this.squareIndex++;
    } else {
      this.tileIndex++;
      if (this.tileIndex >= track.getTrackLength()) {
        this.lap++;
        this.tileIndex = 0;
      }
      const nextTile = track.getTile(this.tileIndex);
      const nextLane = nextTile.laneSquares[this.laneIndex];
      if (!nextLane || nextLane.length === 0) {
        console.error(`Error: Lane ${this.laneIndex} does not exist in tile ${this.tileIndex}!`);
        return;
      }
      this.squareIndex = 0;
      console.log(`${this.name} moved to tile ${this.tileIndex} (${nextTile.color}) in lane ${this.laneIndex}`);
    }
  }

  changeLane(track: Track, direction: number): void {
    const currentTile = track.getTile(this.tileIndex);
    const newLane = this.laneIndex + direction;
    if (newLane >= 0 && newLane <= currentTile.lanes) {
      this.laneIndex = newLane;
      console.log(`${this.name} switched to lane ${this.laneIndex}`);
    } else {
      console.log("Invalid lane change direction!");
    }
  }
}
